src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/DeviceTest.cs

```diff
@@ -74,22 +89,29 @@ protected static bool MonitorAdbLogcat (Func<string, bool> action, string logcat
 			};
 
 			bool didActionSucceed = false;
+			ManualResetEventSlim stdout_done = new ManualResetEventSlim ();
 			using (var sw = File.CreateText (logcatFilePath)) {
 				using (var proc = Process.Start (info)) {
 					proc.OutputDataReceived += (sender, e) => {
 						if (e.Data != null) {
 							sw.WriteLine (e.Data);
 							if (action (e.Data)) {
 								didActionSucceed = true;
-								if (!proc.HasExited) {
-									proc.Kill ();
-								}
 							}
+						} else {
+							stdout_done.Set ();
 						}
 					};
-
 					proc.BeginOutputReadLine ();
-					proc.WaitForExit (timeout * 1000);
+					TimeSpan time = TimeSpan.FromSeconds (timeout);
+					while (!stdout_done.IsSet && !didActionSucceed && time.TotalMilliseconds > 0) {
+						proc.WaitForExit (100);
+						time -= TimeSpan.FromMilliseconds (100);
+					}
+					proc.Kill ();
+					stdout_done.Wait ();
+					proc.WaitForExit ();
```

Comment:
This line needs to be swapped with the previous one - buffers must be flushed after the process exits.

Reply:
Fixed